doc/dev/schema.md

```diff
@@ -0,0 +1,193 @@
+  - [Alternative approaches](#alternative-approaches)
+- [Sharding schema](#sharding-schema)
+  - [How schema is stored](#how-schema-is-stored-1)
+  - [When schema is used](#when-schema-is-used-1)
+  - [How schema is reloaded](#how-schema-is-reloaded-1)
+  - [When schema is reloaded and operation is retried](#when-schema-is-reloaded-and-operation-is-retried-1)
+  - [Alternative approaches](#alternative-approaches-1)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Space schema
+
+Related links: [#98](https://github.com/tarantool/crud/issues/98),
+[PR#111](https://github.com/tarantool/crud/pull/111).
+
+### How schema is stored
+
+Every ``crud`` router is a [``vshard``](https://www.tarantool.io/en/doc/latest/reference/reference_rock/vshard/)
+router, the same applies to storages. In ``vshard`` clusters, spaces are
+created on storages. Thus, each storage has a schema (space list,
+space formats and indexes) on it.
+
+Every router has a [``net.box``](https://www.tarantool.io/en/doc/latest/reference/reference_lua/net_box/)
+connection to each storage it could interact with. (They can be
+retrieved with [``vshard.router.routeall``](https://www.tarantool.io/en/doc/latest/reference/reference_rock/vshard/vshard_router/#lua-function.vshard.router.routeall)
+call.) Each ``net.box`` connection has space schema for an instance it
+is connected to. Router can access space schema by using ``net.box``
+connection object contents.
+
+### When schema is used
+
+Space schema is used to flatten maps on `crud.*_object*` requests.
+Space schema is used to resolve update operation fields if updating
+by field is not supported natively. Together with sharding info,
+the space schema is used to calculate ``bucket_id`` to choose
+replicaset for a request. Finally, schema is returned to a user
+in `metadata` field of request response, so it could be used later
+for `crud.unflatten_rows`.
+
+### How schema is reloaded
+
+``net.box`` schema reload works as follows. For each request (we use ``call``s
+to execute our procedures) from client (router) to server (storage) we
+receive a schema version together with response data. If schema versions mismatch,
+client reloads schema. The request is not retried. ``net.box`` reloads
+schema before returning ``call`` result to a user, so the next request will use
+updated schema. (See [tarantool/tarantool/6169](https://github.com/tarantool/tarantool/issues/6169)).
+``crud`` cannot implicitly rely on ``net.box`` reloads: ``crud`` requests
+use ``net.box`` spaces schema to build ``net.box`` ``call`` request data,
+so if something is not relevant anymore, a part of this request data need
+to be recalculated. ``crud`` uses connection ``reload_schema`` handle
+(``ping`` wrapper, see[PR#111 comment](https://github.com/tarantool/crud/pull/111#issuecomment-765811556))
+to reload schema explicitly when needed. ``crud`` reloads each replicaset
```

Comment:
Don't know, whether it should be written here or it is out of scope of this document, but we can add that `reload_schema()` always sends a ping request (and waits for a response), but not necessarily reloads the schema (content of `_vspace`, `_vindex` and `vcollation` system views). It may give a good guess about a cost of this function.

Reply:
But we rely on `reload_schema` to actually reload schema -- there are no explicit checks on `_vspace` copy and etc.

Reply:
The question is not about that. I'll remove extra information from it:

`reload_schema()` always sends a ping request and waits for a response, but not necessarily reloads the schema. Should we reflect this fact it is the document?

Reply:
Isn't `net.box` developers designed it in such a way that `reload_schema` reload the schema as it is stated by its name?

Reply:
Reloads if necessary. If the schema is actual, it'll not be reloaded.

Reply:
I've tried to improve the text based on this discussion.